### PR TITLE
Add deprecation warning to mysql2 adapter prepared_statements default

### PR DESCRIPTION
**Pull request by thdaraujo**
### Summary

This is the first step into fixing [prepare statements is not enabled by default with mysql2 adapter](https://github.com/rails/rails/issues/42975).

This PR adds a deprecation warning that is raised when `prepared_statements` is not set on the database configuration for the mysql2 adapter.

## Related Pull Request
We are changing the default from `false` to `true`  here: #43238 .

### Questions && Comments
- [x] Should we change the default configuration for tests on ActiveRecord and set the default value to false? The reason for that is that currently we get lots of warnings when running the test suite with mysql2 adapter.
- [x] What message should we display in the deprecation warning?
- [x] We will update the changelog after we get a review to avoid rebasing the PR all the time.
- [x] We noticed there is no documentation regarding the defaults for prepared statements. What do you think of adding this next as well?

### CODE REVIEW

File: activerecord/lib/active_record/database_configurations.rb [hunk 1]
```diff
@@ -70,6 +74,16 @@ def configs_for(env_name: nil, spec_name: nil, name: nil, include_replicas: fals
       end
     end
 
+    def check_prepared_statements_deprecation(config)
+      adapter = config.adapter.to_sym
+
+      if adapter == :mysql2 && !config.configuration_hash.key?(:prepared_statements)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/database_configurations.rb:80]: Why not check this in the mysql2 adapter where we set the value?

Comment (thdaraujo) [reply activerecord/lib/active_record/database_configurations.rb]: Thanks for the review, Rafael. We moved it to the mysql2 adapter 👍

**Review by marcelolx — APPROVED** [review]
> Should we change the default configuration for tests on ActiveRecord and set the default value to false? The reason for that is that currently we get lots of warnings when running the test suite with mysql2 adapter.

I'm not sure but probably yes, the tests are like people's applications, so when they will see the warning they will set the value to false if they want to keep it as it is, with Rails tests I think that it should be the same.

>  What message should we display in the deprecation warning?

I think that the current message is fine.

> We noticed there is no documentation regarding the defaults for prepared statements. What do you think of adding this next as well?

+1

**Comment by thdaraujo** [comment]
Updated the changelog, should be good to go!
Let us know if we need to do anything else here.

**Comment by thdaraujo** [comment]
We are changing the default from false to true here: #43238 .

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb [hunk 2]
```diff
@@ -140,6 +141,14 @@ def discard! # :nodoc:
       end
 
       private
+        def check_prepared_statements_deprecation(config)
+          if !config.key?(:prepared_statements)
+            ActiveSupport::Deprecation.warn(<<-MSG.squish)
+              The default value of `prepared_statements` for the mysql2 adapter will be changed from +false+ to +true+ in Rails 7.1.
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb:147]:
```suggestion
              The default value of `prepared_statements` for the mysql2 adapter will be changed from +false+ to +true+ in Rails 7.2.
```

**Review by rafaelfranca — COMMENTED** [review]
Can you rebase this PR? It looks good to me.

File: activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb (hunk 2 above)
Comment (thdaraujo) [reply activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb]: done, thanks!

**Comment by rafaelfranca** [comment]
Merged in 75fea0bd9590c3ea9447542c5829cf43fe874374

**Pull request closed without merging**